src/librbd/cache/ReplicatedWriteLog.cc

```diff
@@ -513,9 +570,733 @@ template <typename I>
+          front_req = nullptr;
+          ceph_assert(!cleared_dispatching_flag);
+          m_dispatching_deferred_ops = false;
+          cleared_dispatching_flag = true;
+        } else {
+          ceph_assert(!front_req);
+          if (m_deferred_ios.size()) {
+            /* New allocation candidate */
+            front_req = m_deferred_ios.front();
+          } else {
+            ceph_assert(!cleared_dispatching_flag);
+            m_dispatching_deferred_ops = false;
+            cleared_dispatching_flag = true;
+          }
+        }
+      }
+      /* Try allocating for front_req before we decide what to do with allocated_req
+       * (if any) */
+      if (front_req) {
+        ceph_assert(!cleared_dispatching_flag);
+        allocated = front_req->alloc_resources();
+      }
+      if (allocated_req && front_req && allocated) {
+        /* Push dispatch of the first allocated req to a wq */
+        m_work_queue.queue(new LambdaContext(
+          [this, allocated_req](int r) {
+            allocated_req->dispatch();
+          }), 0);
+        allocated_req = nullptr;
+      }
+      ceph_assert(!(allocated_req && front_req && allocated));
+
+      /* Continue while we're still considering the front of the deferred ops list */
+    } while (front_req);
+    ceph_assert(!allocated);
+  }
+  ceph_assert(cleared_dispatching_flag);
+
+  /* If any deferred requests were allocated, the last one will still be in allocated_req */
+  if (allocated_req) {
+    allocated_req->dispatch();
+  }
+}
+
+/**
+ * Returns the lanes used by this write, and attempts to dispatch the next
+ * deferred write
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::release_write_lanes(C_WriteRequestT *write_req)
+{
+  {
+    std::lock_guard locker(m_lock);
+    ceph_assert(write_req->resources.allocated);
+    m_free_lanes += write_req->image_extents.size();
+    write_req->resources.allocated = false;
+  }
+  dispatch_deferred_writes();
+}
+
+/**
+ * Attempts to allocate log resources for a write. Write is dispatched if
+ * resources are available, or queued if they aren't.
+ */
+template <typename I>
+void ReplicatedWriteLog<I>::alloc_and_dispatch_io_req(C_BlockIORequestT *req)
+{
+  bool dispatch_here = false;
+
+  {
+    /* If there are already deferred writes, queue behind them for resources */
+    {
+      std::lock_guard locker(m_lock);
+      dispatch_here = m_deferred_ios.empty();
+    }
+    if (dispatch_here) {
+      dispatch_here = req->alloc_resources();
+    }
+    if (dispatch_here) {
+      if (RWL_VERBOSE_LOGGING) {
+        ldout(m_image_ctx.cct, 20) << "dispatching" << dendl;
+      }
+      req->dispatch();
+    } else {
+      req->deferred();
+      {
+        std::lock_guard locker(m_lock);
+        m_deferred_ios.push_back(req);
+      }
+      if (RWL_VERBOSE_LOGGING) {
+        ldout(m_image_ctx.cct, 20) << "deferred IOs: " << m_deferred_ios.size() << dendl;
+      }
+      dispatch_deferred_writes();
```

Comment:
Nit: should this really expect to be able to dispatch anything here? or should deferred writes be dispatched upon IO completions to the pmem device/flush to the OSD?

Reply:
I agree with you in the theory. This should be called when some resource is freed. Let us do more check.

Reply:
We have to call it. For example, when we checking m_deferred_ios, it is not empty. After the check in Line 771 and before it is put into deferred_ios, dispatch_deferred_writes is triggered and handles all the deferred ios. At this time the resource may be enough but not trigger the dispatch.

Reply:
Line 771?

Reply:
I mean these lines https://github.com/ceph/ceph/pull/29078/files#diff-c01327ef02947ca979424bf5823abc22R769-R770

Reply:
Meanwhile, it is the similar situation when alloca_resources fails for a write request. We don't get ReplicatedWriteLog m_lock all the time. It is possible that the allocation fails but new resources are released before the write request is added into the m_deferred_ios queue. https://github.com/ceph/ceph/blob/edb10684a5521e94ad707fa9289e12ceb33d8459/src/librbd/cache/ReplicatedWriteLog.cc#L773